### PR TITLE
Add from_markdown method

### PR DESCRIPTION
**Pull request by blackary**

### CODE REVIEW

File: stoc.py [hunk 1]
```diff
@@ -9,23 +9,24 @@
...
-        st.write(f"# {text}")
+    def h1(self, text: str, write: bool = True):
+        if write:
+            st.write(f"# {text}")
         self.toc_items.append(("h1", text))
 
-
-    def h2(self, text):
-        st.write(f"## {text}")
+    def h2(self, text: str, write: bool = True):
```
Comment (arnaudmiribel) [inline stoc.py:22]: Thanks for type hints!

File: stoc.py [hunk 2]
```diff
@@ -34,9 +35,27 @@ def toc(self):
         markdown_toc = ""
         for title_size, title in self.toc_items:
             h = int(title_size.replace("h", ""))
-            markdown_toc += " " * 2 * h + "- " + f'<a href="#{normalize(title)}" class="toc"> {title}</a> \n'
+            markdown_toc += (
+                " " * 2 * h
+                + "- "
+                + f'<a href="#{normalize(title)}" class="toc"> {title}</a> \n'
+            )
```
Comment (arnaudmiribel) [inline stoc.py:42]: Better! Some day we'll just use https://github.com/tvst/htbuilder everywhere instead of these raw strings

File: stoc.py [hunk 3]
```diff
@@ -34,9 +35,27 @@ def toc(self):
...
         st.sidebar.write(markdown_toc, unsafe_allow_html=True)
 
+    @classmethod
+    def from_markdown(cls, text: str):
+        self = cls()
+        for line in text.splitlines():
+            if line.startswith("###"):
```
Comment (arnaudmiribel) [inline stoc.py:49]: 🧠

File: streamlit_app.py [hunk 4]
```diff
@@ -13,30 +19,69 @@
...
-st.code("""
-from stoc import stoc
-toc = stoc()
+method1, method2 = st.tabs(["Interspersed with other code", "Block of markdown"])
 
-toc.h1("Demo")
-st.write("...")
+with method1:
+    st.code(
+        dedent(
```
Comment (arnaudmiribel) [inline streamlit_app.py:26]: 🧠 🧠 🧠

Comment (arnaudmiribel) [reply streamlit_app.py:26]: textwrap is such a nice thingy!

**Review by arnaudmiribel — APPROVED** [review]
LGTM! Thank you. Neat idea. Might need to reboot the Cloud app so `streamlit` gets upgraded and supports `st.tabs`

File: stoc.py (hunk 2 above)
Comment (blackary) [reply stoc.py:42]: Or we could use https://pypi.org/project/py-emmet/

**Comment by blackary** [comment]
> LGTM! Thank you. Neat idea. Might need to reboot the Cloud app so `streamlit` gets upgraded and supports `st.tabs`

Shouldn't have to, since I changed requirements.txt to have a new minimum version. We'll see.

**Pull request merged**